src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,337 @@
+            logger.info(msg)
+            modify = True
+
+    # persist an up-to-date version of the pileup data structure in MongoDB
+    if modify:
+        logger.info("monitoring task update %s", doc)
+        mgr.updatePileup(doc)
+
+
+async def inactiveTask(doc, spec):
+    """
+    Perform single inactive task over provided MSPileup document
+    :param mgr: MSPileupData manager
+    :param doc: MSPileup document
+    :param spec: task spec dict
+    """
+    mgr = spec['manager']
+    rucioClient = spec['rucioClient']
+    logger = spec.get('logger', getMSLogger(False))
+    pname = doc['pileupName']
+    inputArgs = {"scope": "cms", "name": pname}
+    ruleIds = [r for r in rucioClient.list_replication_rules(inputArgs)]
+
+    for rdoc in list(ruleIds):
+        # make a Rucio call to delete that rule id
+        rid = rdoc['id']
+        rucioClient.delete_replication_rule(rid)
+
+        # remove the rule id from rulesList (if any) and remove the RSE name from currentRSEs (if any)
+        rse = rdoc['rse_expression']
+        rses = doc['currentRSEs']
+        if rid in ruleIds:
+            ruleIds.remove(rid)
+        if rse in rses:
+            rses.remove(rse)
+            doc['currentRSEs'] = rses
+
+    # make a log record if the DID + Rucio account tuple does not have any existent rules
+    if len(ruleIds):
+        account = rucioClient.whoami()
+        msg = f"Container: {pname} for Rucio account {account} does not have any existing rules"
+        logger.info(msg)
+
+        # and set rulesList and currentRSEs to an empty list
+        doc['currentRSEs'] = []
+        doc['rulesList'] = []
+
+        # persist an up-to-date version of the pileup data structure in MongoDB
+        logger.info("monitoring task update %s", doc)
+        mgr.updatePileup(doc)
+
+
+async def activeTask(doc, spec):
+    """
+    Perform single active task over provided MSPileup document
+    :param mgr: MSPileupData manager
+    :param doc: MSPileup document
+    :param spec: task spec dict
+    """
+    mgr = spec['manager']
+    logger = spec.get('logger', getMSLogger(False))
+    rucioClient = spec['rucioClient']
+    rucioAccount = spec['rucioAccount']
+    marginSpace = spec.get('marginSpace', 0)
+
+    # extract relevant part of our pileup document we'll use in our logic
+    expectedRSEs = set(doc['expectedRSEs'])
+    currentRSEs = set(doc['currentRSEs'])
+    pileupSize = doc['pileupSize']
+    pname = doc['pileupName']
+    inputArgs = {"scope": "cms", "name": pname}
+
+    # if expectedRSEs is different than currentRSEs, then further data placement is required
+    # it's possible that data removal is required!
+    if expectedRSEs != currentRSEs:
+        msg = f"expectedRSEs {expectedRSEs} != currentRSEs {currentRSEs}, further data place is required"
+        logger.warning(msg)
+
+        # for each expectedRSEs not in currentRSEs:
+        for rse in set(expectedRSEs).difference(currentRSEs):
```

Comment:
Why? I doubt it is necessary since list_replication_rules API call is done for pileup name/scope/account (there is no rse involved in API call). Please clarify your request.

Reply:
I don't remember why I suggested it either, given that we are already in a branch where expectedRSEs != currentRSEs. Feel free to mark it as resolved.